### DISCUSSION TITLE
Allow users to change the color of their names

### DISCUSSION THREAD

**Discussion by ghost** [element-hq/element-meta#842] [General]
### Description

I think the user should be able to change the color of their own names (which would be enforced serverside, unless another user's client option overrides it), and I also think that users should be able to force a client side color for all names they see. There should also be an option to choose a random color for each user on the client side. Why should this be done?
- being able to change a color will let users express themselves more
- the random color option will help people identify others faster
- being able to change other colors clientside will make it easier to read for people with cheap/bad monitors, with poor vision and/or in bright environments.
- it's also nice for people who simply don't like the color they see.

Thank you for reading, and I hope you consider my suggestion!

Logs being sent: no, it's not needed

### Version information
- Riot-web: 1.5.6
- OS: Debian 10
- (applies to https://riot.im/app/ and Riot-android Ver. 0.9.9 as well)

**Comment by t3chguy** [comment]
> being able to change other colors clientside will make it easier to read for people with cheap/bad monitors, with poor vision and/or in bright environments.

Most colours people will choose will not play well with themes.

This also begins to feature creep, what if someone wants an animated rainbow name?

**Comment by ghost** [comment]
> > being able to change other colors clientside will make it easier to read for people with cheap/bad monitors, with poor vision and/or in bright environments.
> 
> Most colours people will choose will not play well with themes.

You quoted the clientside part of my suggestion. If the colours don't work with their themes, then they can change it themselves.

> This also begins to feature creep, what if someone wants an animated rainbow name?

we politely tell them no.

**Comment by t3chguy** [comment]
Given you want this enforced server side this is in the wrong repo, it should be in matrix-doc and should probably be structured as an MSC, riot is one client of many for the Matrix protocol.

**Comment by ghost** [comment]
> Given you want this enforced server side this is in the wrong repo, it should be in matrix-doc and should probably be structured as an MSC, riot is one client of many for the Matrix protocol.

Good point. Though, can you tell me what an MSC is? I'm not too familiar with this stuff.

**Comment by t3chguy** [comment]
https://matrix.org/docs/spec/proposals

**Comment by ghost** [comment]
thanks, I'll check that out later today.

Sent from ProtonMail mobile

-------- Original Message --------

**Comment by Kommynct** [comment]
You could limit the colors to a selection, and have that selection change based on whether or not the person is using the dark theme. That'd keep it usable, make it a useful nice bit of personalization, and would allow us to accent things based on their choice of color.

**Comment by krisrobinson** [comment]
> You could limit the colors to a selection, and have that selection change based on whether or not the person is using the dark theme. That'd keep it usable, make it a useful nice bit of personalization, and would allow us to accent things based on their choice of color.

Agree. Limit the pallet but give people the options to change. Half our chat right now is orange and the other is purple, if anything, it's just confusing to what it means, are they mods? why are they purple? Oh its random? We can't change it?

**Comment by ElykDeer** [comment]
I'm just gonna throw my thoughts in here; it's hard to distinguish users in a chat with similar length names of the same color at a glance...handles tend to be pretty static for people across platforms, colors serve as a nice way to distinguish.  Would be nice to have.

In the meantime, is there an admin way to do it on the backend? Wouldn't mind changing it manually for a couple of my users atm.

**Comment by t3chguy** [comment]
No, the colour is never stored, it is a function of the hash of the username.

**Comment by ufgkirk** [comment]
+1 on this feature request... I've had several users on my new Matrix server ask me for this feature.

**Comment by immanuelfodor** [comment]
It would be a relief if I could change the color just for myself, or for anyone in my client with or without backup. I wouldn't care what default color they could see then via id hashing, I just don't like being Mr. Pink :(

**Comment by LoneWolfHT** [comment]
I hate purple. I want a greenish name \:P

**Comment by Panquesito7** [comment]
Sounds like a neat idea 👍 I also wanna change my name color; thought this was possible, as others had other colors. 😄

**Comment by vranki** [comment]
I'd vote against this. Users have different themes and the palette used by Element should be compatible with the used theme. Allowing users to set custom colors would allow unusable colors to be chosen and restricting this would make implementation quite complex.

**Reply by makoConstruct** [reply]
Using the nearest color in the theme to the user's chosen color (`user.display_color = theme.colors.argmin((c)=> (user.chosen_color.as_vec3() - c.as_vec3()).magnitude())`) would not actually be complex.

**Comment by Kommynct** [comment]
By this logic, the current system is also flawed because it chooses colors randomly that don't fit with the theme as well.

**Comment by t3chguy** [comment]
The theme specifies the set of allowed colours. You can create a theme with different colour set.

**Comment by Kommynct** [comment]
Allowing you to choose from the presets does not stop you from creating a theme with a different color set. 

Your green doesn't HAVE to be the same as mine, users don't need to be given full hex control over their colors.

**Comment by immanuelfodor** [comment]
That's why I suggested everybody change colors just on the client side for a current session. This way I could get my favorite color just as I want it, and it wouldn't mess with anyone's theme, it's changing my color only for me. This way I could also change any user's color on the client side, they wouldn't notice it. If this setting would be backed up, it could be transportable between sessions, otherwise, I would be hapy to set it up everytime when I log in for myself.

**Comment by Kommynct** [comment]
porque no las dos?

We could have both, it could be set initially by the user, and you could change it permanently for individual users if you wanted. Like nicknames could have color functionality added onto it. That'd be easy to add onto the feature, you could even add defaults incase you don't like people having them. Or you want it to be based on their role in the group.

Would that not be fair?

**Comment by mitchnull** [comment]
I've submitted a PR for element-android for part of this issue (overriding the colors by the user on their local clients): https://github.com/vector-im/element-android/pull/2614
If this is accepted, I'll try to add the same to element-web and element-desktop, too.

**Comment by masscream** [comment]
Besides all of the reasons of a kind "I don't like the color for my eyes or/and with the particular theme", I've been asked by one of the users of my matrix-server instance, if it is possible to change the color because she types a lot with someone else whose color looks (almost) same by look and therefore it makes the conversations kind of "messy". It's an accident because both users have totally different username. I went through the settings of Element for the variety of platforms and for desktops I can clearly see a lot of possibilities for customizing the font, size, etc, so color shouldn't be missing imho. Thanks for the consideration and your work.

**Comment by t3chguy** [comment]
Themes can specify the username colors, e.g https://github.com/aaronraimist/element-themes

**Comment by masscream** [comment]
I've realized, that I've probably put my answer onto the wrong place. I've meant to address it more to Element as a global client, my friends I've mentioned use the android version the most. Never mind, the request has been sent that way. Sorry.

**Comment by valdeniomarinho** [comment]
Please add this feature ASAP!

**Comment by kittykat** [comment]
We could do with a minor change to [this PR](https://github.com/matrix-org/matrix-react-sdk/pull/5626) for this issue to be closed. If anyone is interested in adding a commit to make the final changes, that would be really helpful!

**Comment by amytimed** [comment]
This should really be added, otherwise you're stuck with a color, again and again, forever.

**Comment by turt2live** [comment]
As an update: we are very much looking to do this, particularly considering you can on mobile but not web, however the team is stretched a bit too thin to take it on ourselves at the moment. We would really appreciate a PR, likely building off the [existing work](https://github.com/matrix-org/matrix-react-sdk/pull/5626), which adds the functionality so we can get cross-platform compatibility here.

If folks would like to take this on, please do - there's no need to ask for permission ;)

**Comment by Yersi88** [comment]
I would be very excited to see this feature implemented, I'm particularly looking forward to more advanced color schemes perhaps gradient of two colours or different effects. I don't think there's a reason to limit ourselves to one colour when it comes to names.

**Comment by turt2live** [comment]
There's a point where the usability of the app becomes severely affected - I'd encourage PRs which focus on solid colours for now.

**Comment by ghost** [comment]
Matrix spec issue: https://github.com/matrix-org/matrix-spec/issues/835

**Reply by OdinVex** [reply]
Not sure what this has to do with clients specifying their own colors for others. I'd set color profiles and export/import to my own devices, much like IRC clients have been doing for 32 years. Took them only two years to come up with communicable colors in an RFC spec but clients had client-side color configuration from near day one.

**Reply by OdinVex** [reply]
(I'm under the opinion I should be able to override someone else's color *locally* because I don't like certain colors, and/or for accessibility reasons.) None of this requires RFCs or Matrix-protocol specs so it could be implemented now and can later be meshed to any spec that may or may not ever come.

**Comment by InklingGirl** [comment]
Any update on this?

**Reply by Baertram** [reply]
Yes, please at least consider client color changes possible. I got so many users with similar colors where I'd love to recolor myself and some of them to cleary see at a glance which user is writing. 
Currently I can only "kind of see" that via the chosen user image, and even those often look similar and are hard to distinguish.
Many thanks

**Comment by noraj** [comment]
What I wrote on https://github.com/matrix-org/matrix-spec/issues/835#issuecomment-3382580801

> > Most colours people will choose will not play well with themes 
> 
> Also display name color could be simply overwritten locally on the client. It would be useful for the user to have a unique color easily recognizable while other users still continue to see this user with an automatically defined color on their client.
> 
> So I don't see why this should necessarily be a Matrix spec issue, it could be a Matrix client feature, if it's just a local client overwrite not reflected on other users.
> 
> > de-clashing colors in DMs and small groups
> 
> It's especially why I want to choose a color for my display name in the first place, even on small groups of 4 people, 2 users have the same color including me and another person. It's already bad on Elements Modern message layout, but it's make it especially easy to misread authors when on IRC message layout.

**Reply by noraj** [reply]
I feel like https://github.com/element-hq/element-meta/issues/841 was closed / converted to discussion because it was understood as a server-side request and so redirected to Matrix spec https://github.com/matrix-org/matrix-spec/issues/835. However, there is no need for that and a client-side feature could be easily implemented. User defined color should apply locally and not be enforced on others (e.g. like on discord).

**Comment by OdinVex** [comment]
I was definitely seeking *client*-side control over it. Edit: Other users could *suggest* a default color. Client should always be able to override for accessibility reasons.

**Comment by Wayyllon** [comment]
Bumping, this is a big nice-to-have in terms of personalization.

**Comment by InklingGirl** [comment]
With the fast approaching Discord exodus, now is the time to get it together w/ QoL/features! We don't want Discord users leaving one bad platform for another, when an encrypted, federated, F/LOSS alternative is available, simply because said alternative is dragging its feet on 5-to-10-year-old suggestions & contributions.

**Reply by immanuelfodor** [reply]
1000% agree

**Reply by OdinVex** [reply]
I think there were far more pressing issues to address than color of names, such as client non-compliance which prevents calls from working: https://github.com/element-hq/element-x-android/issues/4528#issuecomment-3909733057

Forgot I was subscribed to this Issue.
